Fix c10::div_floor_floating compile error

int3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115647

Introduced by #113276. I've added a test to catch future regressions.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

int3 committed:
Fix c10::div_floor_floating compile error

Introduced by #113276. I've added a test to catch future regressions.

[ghstack-poisoned]

vfdev-5 commented:
@int3 thanks for the fix! Is it possible to add a smoke test for that in the inductor tests ?

int3 commented:
I don't really think it's necessary... do you have a particular test file in mind that this test would fit into?

desertfire commented on c10/util/generic_math.h:
```diff
@@ -4,6 +4,17 @@
+#include <c10/hip/HIPMathCompat.h>
+#define compat_copysign c10::hip::compat::copysign
+#else
+#include <c10/util/copysign.h>
```
This worries me. It includes quite a few headers as a closure, which is a threat to the ABI-compatibility mode. You either need to audit all those headers, or you can separate compat_copysign-related code and move it here.

int3 replied:
Isn't this already covered by the fact that we don't link against libc10 in abi-compat mode?

chenyang78 replied:
> Likey true for OSS, but is internal libomnibus going to hide the problem? cc @chenyang78

What ABI-compatibility tries to solve is to make any ABI-related change transparent to the generated DSO. For example, we don't want to expose or passing around c10 tensor structs across the cshim, because any changes to c10 tensor might cause mal-behaviors due to mismatches between DSO and libc10. It doesn't really matter if we link against libc10 or not. 

On the other hand, with abi-compatible cshim, we force all the changes are within libc10 so that we would haven't the aforementioned mismatches. This is why we enforce self-contained header-only files to be included in the generated wrapper code, which won't cause such inconsistency across the cshim boundary.

int3 replied:
I am not sure how libomnibus plays into this. It sounds like it's something that we create in the course of `buck build`; but we aren't using buck to directly compile the `model.so`s, right? I think that's all done through our Python code?

chenyang78 replied:
> I remember during the early AOTI bringup @chenyang78 had to link against libomnibus instead of libtorch. Not sure about the latest status, if @chenyang78 can confirm this is not the case anymore, I am ok with this PR.

Yeah, I don't think libomnibus is an issue for this case.

vfdev-5 commented:
> I don't really think it's necessary... do you have a particular test file in mind that this test would fit into?

You can add the following test into `pytorch/test/inductor/test_torchinductor.py` (for example below: https://github.com/pytorch/pytorch/blob/1500379b6da2bf3122748e0dffdf9a775d117b6b/test/inductor/test_torchinductor.py#L1652)

```python
    def test_floordiv(self):
        def fn(a, i):
            n = (i * 1.0) // 8.0
            return a + n

        self.common(fn, (make_tensor(100, device="cpu", dtype=torch.float32), 12))
```

Right now this test gives 2 failures: 1) on cpu with `/pytorch/torch/include/c10/util/generic_math.h:43:31: error: ‘compat_copysign’ was not declared in this scope  floordiv = compat_copysign(scalar_t(0), a / b);` and 2) on cuda (to be explored why)

peterbell10 commented on c10/util/generic_math.h:
```diff
@@ -4,6 +4,17 @@
 
+#if defined(__CUDA_ARCH__)
+#include <c10/cuda/CUDAMathCompat.h>
+#define compat_copysign c10::cuda::compat::copysign
```
Can we rename to C10_COMPAT_COPYSIGN` to avoid macro pollution issues?

int3 committed:
Update on "Fix c10::div_floor_floating compile error"


Introduced by #113276. I've added a test to catch future regressions.

[ghstack-poisoned]

vfdev-5 reviewed (approved):
Thanks @int3, lgtm

int3 commented:
@pytorchbot merge

int3 commented:
@pytorchbot merge
